### PR TITLE
Fix premature processing for gitea repos

### LINKED ISSUE #100: The processor stuck at `addDir` when processing repos
Sometimes after migrating the first repo to gitea, the processor gets stuck at the `addDir` step. restarting the container fixes the problem.

### PR DESCRIPTION
**Pull request by AbdulrhmnGhanem**
Fixes #100
## 
- When migrating a repo [1]: Gitea creates an empty repo then the migration
  starts.
- When uploading user files [2]: we create an empty repo,
  upload the files, finally commit the files to the repo.
- When the processor gets notified by a file-system change when the
  repository is created in both [1], and [2] the repo is still empty and
  the `git checkout` command fails.
- The fix:
    - only process repos if it isn't empty,
    - only process synced repos when the migration task is done.
 ##
* The processor check gitea database for the migration state of the repo.
* The maximum time for retries can be configured in the evn via `MAXIMUM_REPO_MIGRATION_TIME`.
##
* `e2e` doesn't have any flaky tests now :tada:

### CODE REVIEW

File: processor/src/processingManager.js [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+      .backoff()
+  })
+}
+
+const ProcessingManager = {
+  client,
+  isRepoReadyForProcessing,
+}
+
+module.exports = { ProcessingManager }
```
Comment (kasbah) [inline processor/src/processingManager.js:233]: We need a better name for all of this. Name it after what it's doing, it's definitely not "managing" the processing.

File: processor/src/processingManager.js [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+
+    const onFail = reject =>
+      reject(
+        new Error(
+          `Repo: ${ownerName}/${repoName} not found after ${MaximumRetries} trials.`,
+        ),
+      )
+
+    return asyncBackoff(onBackoff, onFail, MAXIMUM_NUM_OF_RETRIES)
+  }
```
Comment (kasbah) [inline processor/src/processingManager.js:132]: Can this function be defined outside the top level function?

File: processor/src/processingManager.js [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+
+    const onFail = reject =>
+      reject(
+        new Error(
+          `Repo: ${repoId} migration task not found after ${MaximumRetries} trials.`,
+        ),
+      )
+
+    return asyncBackoff(onBackoff, onFail, MAXIMUM_NUM_OF_RETRIES)
+  }
```
Comment (kasbah) [inline processor/src/processingManager.js:181]: Here too: can this function be defined outside the top level function?

File: processor/src/processingManager.js (hunk 2 above)
Comment (kasbah) [reply processor/src/processingManager.js:132]: I'm wondering about the backoff as well: can we use a Postgres/SQL lib that would take care of the lowere level details?

File: processor/src/app.js [hunk 4]
```diff
@@ -2,15 +2,15 @@ const express = require('express')
 
 const { createProjectsAPI } = require('./projectAPI')
 const { createRemoteAPI } = require('./remoteAPI')
-const { ProcessingManager } = require('./processingManager')
+const { isRepoReadyForProcessing } = require('./isReadyForProcessing')
```
Comment (kasbah) [inline processor/src/app.js:5]: I generally try and follow this rule: `is..` is a boolean variable, never a function. So I would suggeest: `checkIsRepoReady` and call the file `checkIsRepoReady`.

File: processor/src/processingManager.js (hunk 1 above)
Comment (AbdulrhmnGhanem) [reply processor/src/processingManager.js]: Renamed it to `isReadyForPorcessing`.

File: processor/test/integration/test_projects.js [hunk 5]
```diff
@@ -31,7 +31,7 @@ describe('projects API', () => {
   beforeEach(async () => {
     await exec(`mkdir -p ${tmpDir}`)
     await exec(`mkdir -p ${repoDir}`)
-    this.app = createApp(repoDir, processingManagerStub)
+    this.app = createApp(repoDir, isRepoReadyForProcessingStub)
     this.supertest = supertest(this.app)
   })
```
Comment (kasbah) [inline processor/test/integration/test_projects.js:37]: That stub is really ugly, can you not let it be `undefined` and check it down the line?

File: processor/src/watcher.js [hunk 6]
```diff
@@ -24,7 +33,7 @@ function watch(events, repoDir = '/repositories') {
     // additionally we ignore any invocations that happen while it's already running
     // to prevent it from trying to overwrite files that are already being written to
     const debouncedProcessRepo = debounce(async () => {
-      if (!running[gitDir]) {
+      if ((await isRepoReadyForProcessing(gitDir)) && !running[gitDir]) {
```
Comment (kasbah) [inline processor/src/watcher.js:36]:
```suggestion
const isReady =  isRepoReadyForProcessing == null || (await isRepoReadyForProcessing(gitDir))
if (!running[gitDir] && isReady) {
```

File: processor/src/watcher.js [hunk 7]
```diff
@@ -13,7 +14,15 @@ const processKicadPCB = require('./tasks/processKicadPCB')
 const processReadme = require('./tasks/processReadme')
 
 const running = {}
-function watch(events, repoDir = '/repositories') {
+
+/**
+ *
+ * @param {*} events
+ * @param {string=} repoDir
+ * @param {checkIsRepoReady} checkIsRepoReady
```
Comment (kasbah) [inline processor/src/watcher.js:22]:
```suggestion
 * @param {function=} checkIsRepoReady
```

**Pull request merged**